### ISSUE TITLE
randomize_layout attribute is not documented

### ISSUE THREAD

**Issue by pinskia** [llvm/llvm-project#111320]
There is no documentation for the randomize_layout attribute in the clang documentation.

**Comment by pinskia** [comment]
Actualy it is documented:
https://clang.llvm.org/docs/AttributeReference.html#randomize-layout-no-randomize-layout

But it does not document how :
"The randomization is deterministic based for a given seed, so the entire program should be compiled with the same seed, but keep the seed safe otherwise. " works.

Is it based on the struct name? Or something else? Or the order which the struct is defined? Or is the seed the same for each struct? It is not obvious?

Does clang record the seed and then error out if there are 2 TUs with the same seed?
Seems like if this is a security feature then it should warn at least?

**Comment by pinskia** [comment]
Does clang warn about when using randomize_layout with pointers? seems like without this is a security issue waiting to happen.

**Comment by AaronBallman** [comment]
Yeah, this documentation should be updated. For example, it turns out this randomization happens automatically if the structure (not union) is not marked `no_randomize_layout`, contains only function pointer members, and a seed is supplied on the command line. Surprise! https://godbolt.org/z/zdnEYbe7b

Also, it seems the documentation is duplicated; I see two entries for these attributes:
https://clang.llvm.org/docs/AttributeReference.html#randomize-layout-no-randomize-layout
https://clang.llvm.org/docs/AttributeReference.html#id10

> Is it based on the struct name? Or something else? Or the order which the struct is defined? Or is the seed the same for each struct? It is not obvious?

It's based on the supplied seed from the command line combined with the name of the structure declaration.

> Does clang record the seed and then error out if there are 2 TUs with the same seed?

I don't believe so.

> Does clang warn about when using randomize_layout with pointers?

Making sure I understand the question, do you mean when passing a pointer to a randomize_layout object? Or do you mean something else?

CC @connorkuehl @isanbard
